### PR TITLE
Filter_block

### PR DESCRIPTION
**Pull request by zalmane**

### CODE REVIEW

File: datayoga/blocks/expression.py [hunk 1]
```diff
@@ -32,12 +32,41 @@ def search(self, data: Any) -> Any:
...
+            boolean: True if matches, else False
+        """
+        clauses = []
+        for k, v in data.items():
+            clauses.append(f"{json.dumps(v)} as '{k}'")
+
+        from_clause = f"select {','.join(clauses)}"
+        print(f"select 1 from ({from_clause}) where {self.expression}")
+        print(self.conn.execute(f"select 1 from ({from_clause}) where {self.expression}").fetchone())
+        return self.conn.execute(f"select 1 from ({from_clause}) where {self.expression}").fetchone() is not None
```
Comment (spicy-sauce) [inline datayoga/blocks/expression.py]: you can use a variable instead of repeating it...

File: datayoga/blocks/filter/block.schema.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{
+  "description": "Filter records",
+  "title": "Filter records",
```
Comment (spicy-sauce) [inline datayoga/blocks/filter/block.schema.json:2]: let's have the title first.

File: datayoga/blocks/filter/tests/test_filter.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+from datayoga.blocks.filter.block import Block
+
+
+def test_filter_sql():
+    block = Block({"language": "sql",
+                   "expression": "age>20"})
```
Comment (spicy-sauce) [inline datayoga/blocks/filter/tests/test_filter.py]: can be in one line or the assert should be in the same format (field under field), same for the rest of the tests.

**Review by spicy-sauce — CHANGES_REQUESTED** [review]

File: datayoga/blocks/expression.py (hunk 1 above)
Comment (zalmane) [reply datayoga/blocks/expression.py]: refactored

File: datayoga/blocks/filter/block.schema.json (hunk 2 above)
Comment (zalmane) [reply datayoga/blocks/filter/block.schema.json:2]: done

File: datayoga/block.py [hunk 4]
```diff
@@ -46,7 +46,7 @@ def init(self):
         """
         pass
 
-    def transform(self, data: Any, context: Context = None) -> Any:
+    def transform(self, data: List[Dict[str, Any]], context: Context = None) -> Any:
```
Comment (spicy-sauce) [inline datayoga/block.py]: shouldn't it also return List[Dict[str, Any]]?

File: datayoga/blocks/expression.py [hunk 5]
```diff
@@ -32,12 +32,58 @@ def search(self, data: Any) -> Any:
...
+        """
+        pass
+
 
 class SQLExpression(Expression):
     def compile(self, expression: str):
         self.conn = sqlite3.connect(":memory")
         self.expression = expression
 
+    def filter(self, data: List[Any]) -> List[Any]:
```
Comment (spicy-sauce) [inline datayoga/blocks/expression.py]: should be changed to List[Dict[str, Any]] (both places)

File: datayoga/blocks/expression.py [hunk 6]
```diff
@@ -32,12 +32,58 @@ def search(self, data: Any) -> Any:
...
+    def test(self, data: Any) -> bool:
+        """Test a where clause for an SQL statement
+
+        Args:
+            data (Any): Data
+
+        Returns:
+            boolean: True if matches, else False
+        """
+        pass
```
Comment (spicy-sauce) [inline datayoga/blocks/expression.py]: why not implemented?

File: datayoga/blocks/expression.py [hunk 7]
```diff
@@ -32,12 +32,58 @@ def search(self, data: Any) -> Any:
...
+        columns_clause = ','.join(f"`{col}`" for col in column_names)
+
+        # values in the form of (?,?), (?,?)
+        values_clause_row = f"({','.join('?'*len(column_names))})"
+        values_clause = ','.join([values_clause_row]*len(data))
+
+        # use a CTE to create the in memory data structure
+        return f"with data({columns_clause}) as (values {values_clause})"
+
     def search(self, data: Any) -> Any:
```
Comment (spicy-sauce) [inline datayoga/blocks/expression.py]: should be List[Dict[str, Any]] (both places)

File: datayoga/blocks/filter/block.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+logger = logging.getLogger(__name__)
+
+
+class Block(DyBlock):
+    def init(self):
+        logger.debug(f"Initializing {self.get_block_name()}")
+        self.language = self.properties["language"]
+        self.expression = expression.compile(self.language, self.properties["expression"])
+
+    def run(self, data: Any, context: Context = None) -> Any:
```
Comment (spicy-sauce) [inline datayoga/blocks/filter/block.py]: change Any to  List[Dict[str, Any]]

File: datayoga/blocks/filter/tests/test_filter.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
+from datayoga.blocks.filter.block import Block
+
+
```
Comment (spicy-sauce) [inline datayoga/blocks/filter/tests/test_filter.py:3]: add a test on a bunch of records...

File: datayoga/blocks/map/block.py [hunk 10]
```diff
@@ -12,10 +12,15 @@
...
+        self.expression = expression.compile(self.properties["language"], json.dumps(self.properties["expression"]))
 
     def run(self, data: Any, context: Context = None) -> Any:
         logger.debug(f"Running {self.get_block_name()}")
-        data = self.expression.search(data)
+        return_data = []
+        for row in data:
+            print(row)
+            print(self.properties)
+            print(self.expression)
```
Comment (spicy-sauce) [inline datayoga/blocks/map/block.py]: remove

File: datayoga/blocks/remove_field/block.py [hunk 11]
```diff
@@ -14,5 +14,6 @@ def init(self):
     def run(self, data: Any, context: Context = None) -> Any:
```
Comment (spicy-sauce) [inline datayoga/blocks/remove_field/block.py]: Change Any to  List[Dict[str, Any]]

File: datayoga/blocks/rename_field/block.py [hunk 12]
```diff
@@ -11,8 +11,8 @@ class Block(DyBlock):
     def init(self):
         logger.debug(f"Initializing {self.get_block_name()}")
 
-    def run(self, data: Any, context: Context = None) -> Any:
+    def run(self, data: List[Dict[str, Any]], context: Context = None) -> Any:
```
Comment (spicy-sauce) [inline datayoga/blocks/rename_field/block.py]: return type should also be List[Dict[str, Any]]

File: datayoga/blocks/remove_field/block.py [hunk 13]
```diff
@@ -14,5 +14,6 @@ def init(self):
...
         logger.debug(f"Running {self.get_block_name()}")
 
-        del data[self.properties["field"]]
+        for row in data:
+            del row[self.properties["field"]]
```
Comment (spicy-sauce) [inline datayoga/blocks/remove_field/block.py:18]: should we manipulate the data itself or cloning it first?

**Review by spicy-sauce — CHANGES_REQUESTED** [review]

File: datayoga/block.py (hunk 4 above)
Comment (zalmane) [reply datayoga/block.py]: yes. changed.

File: datayoga/blocks/expression.py (hunk 7 above)
Comment (zalmane) [reply datayoga/blocks/expression.py]: done

File: datayoga/blocks/expression.py (hunk 5 above)
Comment (zalmane) [reply datayoga/blocks/expression.py]: done

File: datayoga/blocks/expression.py (hunk 6 above)
Comment (zalmane) [reply datayoga/blocks/expression.py]: removed. in the beginning I looped and ran 'test' but more efficient to just filter

File: datayoga/blocks/filter/tests/test_filter.py (hunk 9 above)
Comment (zalmane) [reply datayoga/blocks/filter/tests/test_filter.py:3]: done

File: datayoga/blocks/map/block.py (hunk 10 above)
Comment (zalmane) [reply datayoga/blocks/map/block.py]: removed

File: datayoga/blocks/remove_field/block.py (hunk 11 above)
Comment (zalmane) [reply datayoga/blocks/remove_field/block.py]: done

File: datayoga/blocks/rename_field/block.py (hunk 12 above)
Comment (zalmane) [reply datayoga/blocks/rename_field/block.py]: done

File: datayoga/blocks/remove_field/block.py (hunk 13 above)
Comment (zalmane) [reply datayoga/blocks/remove_field/block.py:18]: good question. maybe this should happen outside the block so it is generic? only downside is the performance penalty

File: datayoga/blocks/add_field/block.py [hunk 14]
```diff
@@ -11,11 +11,12 @@
 class Block(DyBlock):
     def init(self):
         logger.debug(f"Initializing {self.get_block_name()}")
-        self.expression = get_expression_class(self.properties["language"], self.properties["expression"])
+        self.expression = expression.compile(self.properties["language"], self.properties["expression"])
 
     def run(self, data: Any, context: Context = None) -> Any:
```
Comment (spicy-sauce) [inline datayoga/blocks/add_field/block.py]: data: Any should be replaced with data: List[Dict[str, Any]] and return type from Any to List[Dict[str, Any]].

File: datayoga/blocks/expression.py [hunk 15]
```diff
@@ -38,7 +38,42 @@ def compile(self, expression: str):
         self.conn = sqlite3.connect(":memory")
         self.expression = expression
 
-    def search(self, data: Any) -> Any:
+    def filter(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+        """Test a where clause for an SQL statement
```
Comment (spicy-sauce) [inline datayoga/blocks/expression.py]: Test --> Tests

File: datayoga/blocks/expression.py [hunk 16]
```diff
@@ -57,24 +92,35 @@ def exec_sql(self, data: Any, expression: str) -> Any:
...
+
+        return cursor.fetchone()[0]
 
 
 class JMESPathExpression(Expression):
     def compile(self, expression: str):
         self.expression = jmespath.compile(expression)
+        self.filter_expression = jmespath.compile(f"[?{expression}]")
+
+    def filter(self, data: Any) -> bool:
```
Comment (spicy-sauce) [inline datayoga/blocks/expression.py]: data: Any should be replaced with data: List[Dict[str, Any]]

File: datayoga/blocks/expression.py [hunk 17]
```diff
@@ -57,24 +92,35 @@ def exec_sql(self, data: Any, expression: str) -> Any:
...
+        return self.filter_expression.search(data)
 
     def search(self, data: Any) -> Any:
```
Comment (spicy-sauce) [inline datayoga/blocks/expression.py]: data: Any should be replaced with data: List[Dict[str, Any]] and return type from Any to List[Dict[str, Any]].

File: datayoga/blocks/filter/block.py (hunk 8 above)
Comment (spicy-sauce) [inline datayoga/blocks/filter/block.py]: data: Any should be replaced with data: List[Dict[str, Any]] and return type from Any to List[Dict[str, Any]].

File: datayoga/blocks/map/block.py [hunk 18]
```diff
@@ -12,10 +12,12 @@
 class Block(DyBlock):
     def init(self):
         logger.debug(f"Initializing {self.get_block_name()}")
-        self.expression = get_expression_class(self.properties["language"], json.dumps(self.properties["expression"]))
+        self.expression = expression.compile(self.properties["language"], json.dumps(self.properties["expression"]))
 
     def run(self, data: Any, context: Context = None) -> Any:
```
Comment (spicy-sauce) [inline datayoga/blocks/map/block.py:17]: data: Any should be replaced with data: List[Dict[str, Any]] and return type from Any to List[Dict[str, Any]].

File: datayoga/blocks/remove_field/block.py (hunk 13 above)
Comment (spicy-sauce) [reply datayoga/blocks/remove_field/block.py:18]: yes, probably outside of the block is the right approach.

**Review by spicy-sauce — CHANGES_REQUESTED** [review]

File: datayoga/blocks/add_field/block.py (hunk 14 above)
Comment (zalmane) [reply datayoga/blocks/add_field/block.py]: done

File: datayoga/blocks/expression.py (hunk 16 above)
Comment (zalmane) [reply datayoga/blocks/expression.py]: done

File: datayoga/blocks/expression.py (hunk 15 above)
Comment (zalmane) [reply datayoga/blocks/expression.py]: done

File: datayoga/blocks/expression.py (hunk 17 above)
Comment (zalmane) [reply datayoga/blocks/expression.py]: done

File: datayoga/blocks/filter/block.py (hunk 8 above)
Comment (zalmane) [reply datayoga/blocks/filter/block.py]: done

File: datayoga/blocks/remove_field/block.py (hunk 13 above)
Comment (zalmane) [reply datayoga/blocks/remove_field/block.py:18]: let's open a separate ticket to do it there

Comment (spicy-sauce) [reply datayoga/blocks/remove_field/block.py:18]: https://github.com/datayoga-io/datayoga-py/issues/33

**Review by spicy-sauce — APPROVED** [review]

**Pull request merged**